gaming/src/golang/item-service/main_test.go

```diff
@@ -0,0 +1,450 @@
+	}
+	serviceContainer, err := testcontainers.GenericContainer(ctx, testcontainers.GenericContainerRequest{
+		ContainerRequest: req,
+		Started:          true,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Retrieve the container endpoint
+	endpoint, err := serviceContainer.Endpoint(ctx, "")
+	if err != nil {
+		return nil, err
+	}
+
+	return &Service{
+		Container: serviceContainer,
+		Endpoint:  endpoint,
+	}, nil
+}
+
+func TestMain(m *testing.M) {
+	ctx := context.Background()
+
+	// Setup the docker network so containers can talk to each other
+	nr := testcontainers.NetworkRequest{
+		Name:       TESTNETWORK,
+		Attachable: true,
+	}
+	_, err := testcontainers.GenericNetwork(ctx, testcontainers.GenericNetworkRequest{
+		NetworkRequest: nr,
+	})
+	if err != nil {
+		fmt.Printf("Error setting up docker test network: %s\n", err)
+		os.Exit(1)
+	}
+
+	// Setup the emulator container and default instance/database
+	spannerEmulator, err := setupSpannerEmulator(ctx)
+	if err != nil {
+		fmt.Printf("Error setting up emulator: %s\n", err)
+		os.Exit(1)
+	}
+
+	// Run service
+	service, err := setupService(ctx, spannerEmulator)
+	if err != nil {
+		fmt.Printf("Error setting up service: %s\n", err)
+		os.Exit(1)
+	}
+
+	defer teardown(ctx, spannerEmulator, service)
+
+	os.Exit(m.Run())
+}
+
+func httpPUT(url string, data io.Reader) (*http.Response, error) {
+	client := &http.Client{}
+	req, err := http.NewRequest(http.MethodPut, url, data)
+	if err != nil {
+		return nil, err
+	}
+	// set the request header Content-Type for json
+	req.Header.Set("Content-Type", "application/json")
+	response, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	return response, nil
+}
+
+func TestCreateGameItems(t *testing.T) {
+	response, err := http.Post("http://localhost/items", "application/json", bytes.NewBuffer([]byte("{\"item_name\": \"test item\",\"item_value\": \"3.14\"}")))
+	if err != nil {
+		panic(err.Error())
+	}
+
+	assert.Equal(t, 201, response.StatusCode)
+}
+
+func TestPlayerAcquireItemsAndMoney(t *testing.T) {
+	// Test getting a player "/players" endpoint
+	response, err := http.Get("http://localhost/players")
+	if err != nil {
+		panic(err.Error())
+	}
+	assert.Equal(t, 200, response.StatusCode)
+
+	body, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		panic(err.Error())
+	}
+
+	var pData models.Player
+	json.Unmarshal(body, &pData)
+
+	assert.NotEmpty(t, pData.PlayerUUID)
+	assert.NotEmpty(t, pData.Current_game)
+
+	// Have the player acquire money, test the '/players/balance' PUT endpoint
+	type PlayerBalanceRequest struct {
+		PlayerUUID, Source, Amount string
+	}
+
+	r := big.NewRat(1, 1)
+	_, ok := r.SetString("4.30")
+	assert.True(t, ok)
+
+	testPB := PlayerBalanceRequest{PlayerUUID: pData.PlayerUUID, Source: "loot", Amount: r.FloatString(2)}
+	pbJSON, _ := json.Marshal(testPB)
+
+	response, err = httpPUT("http://localhost/players/balance", bytes.NewBuffer(pbJSON))
+	if err != nil {
+		panic(err.Error())
```

Comment:
Here and elsewhere: Would it not make more sense to use `t.Error(err)`?

Reply:
This is where my unfamiliarity sets in. I was using `panic` here to not proceed with the test. `t.Error()` seems to [continue execution](https://pkg.go.dev/testing#T.Error) of the test function.

But, it seems `t.Fatal` is what I want. Will update.